task-impls/src/da.rs

```diff
@@ -0,0 +1,334 @@
+        _ => {}
+    }
+    (None, state)
+}
+
+impl<
+        TYPES: NodeType<ConsensusType = SequencingConsensus, Time = ViewNumber>,
+        I: NodeImplementation<
+            TYPES,
+            Leaf = SequencingLeaf<TYPES>,
+            ConsensusMessage = SequencingMessage<TYPES, I>,
+        >,
+    > DATaskState<TYPES, I>
+where
+    I::Exchanges: SequencingExchangesType<TYPES, Message<TYPES, I>>,
+    CommitteeEx<TYPES, I>: ConsensusExchange<
+        TYPES,
+        Message<TYPES, I>,
+        Certificate = DACertificate<TYPES>,
+        Commitment = TYPES::BlockType,
+    >,
+{
+    pub async fn handle_event(
+        &mut self,
+        event: SequencingHotShotEvent<TYPES, I>,
+    ) -> Option<HotShotTaskCompleted> {
+        match event {
+            SequencingHotShotEvent::DAProposalRecv(proposal, sender) => {
+                let view = proposal.data.get_view_number();
+                if view < self.cur_view {
+                    return None;
+                }
+                let block_commitment = proposal.data.deltas.commit();
+                let view_leader_key = self.committee_exchange.get_leader(view);
+                if view_leader_key != sender {
+                    return None;
+                }
+                if !view_leader_key.validate(&proposal.signature, block_commitment.as_ref()) {
+                    warn!(?proposal.signature, "Could not verify proposal.");
+                    return None;
+                }
+
+                let vote_token = self.committee_exchange.make_vote_token(view);
+                match vote_token {
+                    Err(e) => {
+                        error!("Failed to generate vote token for {:?} {:?}", view, e);
+                    }
+                    Ok(None) => {
+                        info!("We were not chosen for DA committee on {:?}", view);
+                    }
+                    Ok(Some(vote_token)) => {
+                        info!("We were chosen for DA committee on {:?}", view);
+
+                        // Generate and send vote
+                        let message = self.committee_exchange.create_da_message::<I>(
+                            self.high_qc.commit(),
+                            block_commitment,
+                            view,
+                            vote_token,
+                        );
+
+                        self.cur_view = view;
+
+                        if let CommitteeConsensusMessage::DAVote(vote) = message {
+                            info!("Sending vote to the DA leader {:?}", vote);
+                            self.event_stream
+                                .publish(SequencingHotShotEvent::DAVoteSend(vote))
+                                .await;
+                        }
+                    }
+                }
+            }
+            SequencingHotShotEvent::DAVoteRecv(vote, sender) => {
+                if vote.signature.0 != <TYPES::SignatureKey as SignatureKey>::to_bytes(&sender) {
```

Comment:
Just like in consensus, we should probably check first that we are the leader for the view of this vote